content/blog/2024/2024-05-15-java-functional-interfaces.md

```diff
@@ -0,0 +1,2028 @@
+
+**Function composition is a process of combining multiple functions to create a new function.** The `compose()` method in `Function` interface combines two functions by applying the argument function first and then the caller function. Conversely, the `andThen()` method applies the caller function first and then the argument function.
+
+For example, if we have two functions: one to convert a string to upper case and another to remove vowels from it, we can compose them using `compose()` or `andThen()`. If we use `compose()`, it first converts the string to uppercase and then removes vowels from it. Conversely, if we use `andThen()`, it first removes vowels from it and then converts the string to uppercase.
+
+Let's verify function composition:
+
+```java
+void functionComposition() {
+  Function<String, String> toUpper = s -> s == null ? null : s.toUpperCase();
+  Function<String, String> replaceVowels =
+      s ->
+          s == null
+              ? null
+              : s.replace("A", "")
+                  .replace("E", "")
+                  .replace("I", "")
+                  .replace("O", "")
+                  .replace("U", "");
+  Assertions.assertEquals("APPLE", toUpper.compose(replaceVowels).apply("apple"));
+  Assertions.assertEquals("PPL", toUpper.andThen(replaceVowels).apply("apple"));
+}
+```
+
+In the `functionComposition` test, we compose two functions to manipulate a string. The first function converts the string to uppercase, while the second one removes vowels. Using `compose()`, it first removes vowels and then converts to uppercase. Using `andThen()`, it first converts to uppercase and then removes vowels. We verify the results using assertion.
+
+### BiFunction
+
+**The `BiFunction` interface represents a function that accepts two arguments and produces a result.** It's similar to the Function interface, but it operates on two input parameters instead of one.
+
+```java
+@FunctionalInterface
+public interface BiFunction<T, U, R> {
+  R apply(T t, U u);
+  // default methods
+```
+
+This is the specialized version of `Function` with two arguments. It is a functional interface that defines the `apply(Object, Object)` functional method.
+
+For example, suppose we have a `BiFunction` that takes two integers as input and returns the bigger number.
+
+Let's define it and test the results:
+
+```java
+
+@Test
+void biFunction() {
+  BiFunction<Integer, Integer, Integer> bigger =
+      (first, second) -> first > second ? first : second;
+  Function<Integer, Integer> square = number -> number * number;
+
+  Assertions.assertEquals(10, bigger.apply(4, 10));
+  Assertions.assertEquals(100, bigger.andThen(square).apply(4, 10));
+}
+```
+
+The `BiFunction` interface combines two input values and produces a result. In this test, `bigger` selects the larger of two integers. `square` then calculates the square of a number. The result of `bigger` is passed to `square`, which squares the larger integer.
+
+### IntFunction
```

Comment:
I think that for functions, operators, consumers, and supplier, it's sufficient to include the integer version and then give a list of the other available variants. Now, the article feels like reading a Java doc that purely lists methods. What do you think?

Reply:
Removed them.